File: src/main/java/io/github/svegon/capi/event/render/EnableAmbientOcclusionCallback.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@FunctionalInterface
public interface EnableAmbientOcclusionCallback {
    Event<EnableAmbientOcclusionCallback> EVENT = EventFactory.createArrayBacked(
            EnableAmbientOcclusionCallback.class, callback -> {}, listeners -> callback -> {
                for (EnableAmbientOcclusionCallback listener : listeners) {
                    listener.isAmbientOcclusionEnabled(callback);

                    if (callback.isCancelled()) {
                        return;
                    }
                }
            });

    void isAmbientOcclusionEnabled(CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/EnableAmbientOcclusionCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.EnableAmbientOcclusionCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

fun interface EnableAmbientOcclusionCallback {
    fun isAmbientOcclusionEnabled(callback: CallbackInfoReturnable<Boolean?>?)

    companion object {
        @JvmField
        val EVENT: Event<EnableAmbientOcclusionCallback> = EventFactory.createArrayBacked(
            EnableAmbientOcclusionCallback::class.java,
            EnableAmbientOcclusionCallback { callback: CallbackInfoReturnable<Boolean?>? -> }) { listeners: Array<EnableAmbientOcclusionCallback> ->
            EnableAmbientOcclusionCallback { callback: CallbackInfoReturnable<Boolean?> ->
                for (listener in listeners) {
                    listener.isAmbientOcclusionEnabled(callback)

                    if (callback.isCancelled) {
                        return@EnableAmbientOcclusionCallback
                    }
                }
            }
        }
    }
}
